tools/state-parts-dump/src/cli.rs

```diff
@@ -0,0 +1,167 @@
+use near_client::sync::external::{
+    external_storage_location, external_storage_location_directory, get_num_parts_from_filename
+};
+extern crate rayon;
+use rayon::prelude::*;
+use near_primitives::state_part::PartId;
+use near_primitives::types::{EpochId, ShardId, StateRoot};
+
+use borsh::BorshDeserialize;
+use near_client::sync::external::{create_bucket_readonly, ExternalConnection};
+use near_store::Trie;
+use std::time::Duration;
+use std::path::PathBuf;
+use std::sync::Arc;
+use anyhow::anyhow;
+
+pub struct StatePartsDumpCheck {
+    chain_id: String,
+    epoch_id: EpochId,
+    epoch_height: u64,
+    shard_id: ShardId,
+    state_root: StateRoot,
+    root_dir: Option<PathBuf>,
+    s3_bucket: Option<String>,
+    s3_region: Option<String>,
+    gcs_bucket: Option<String>,
+}
+
+pub enum StatePartsDumpCheckStatus {
+    Done {
+        epoch_height: u64
+    },
+    WaitingForParts{
+        epoch_height: u64
+    },
+}
+
+impl StatePartsDumpCheck {
+    pub fn new(
+        chain_id: String,
+        epoch_id: EpochId,
+        epoch_height: u64,
+        shard_id: ShardId,
+        state_root: StateRoot,
+        root_dir: Option<PathBuf>,
+        s3_bucket: Option<String>,
+        s3_region: Option<String>,
+        gcs_bucket: Option<String>,
+    ) -> Self {
+        Self {
+            chain_id,
+            epoch_id,
+            epoch_height,
+            shard_id,
+            state_root,
+            root_dir,
+            s3_bucket,
+            s3_region, 
+            gcs_bucket
+        }
+    }
+    
+    pub async fn run(&self) -> anyhow::Result<StatePartsDumpCheckStatus> {
+        crate::metrics::STATE_SYNC_DUMP_CHECK_EPOCH_HEIGHT.with_label_values(&[&self.shard_id.to_string()]).set(self.epoch_height as i64);
+
+        let external = create_external_connection(
+            self.root_dir.clone(),
+            self.s3_bucket.clone(),
+            self.s3_region.clone(),
+            self.gcs_bucket.clone(),
+        );
+
+        let directory_path = external_storage_location_directory(&self.chain_id, &self.epoch_id, self.epoch_height, self.shard_id);
+        let part_file_names = external.list_state_parts(self.shard_id, &directory_path).await?;
+        // TODO: return error instead of assert!()
+        assert!(!part_file_names.is_empty(), "Number of state parts is 0");
+        let num_parts = part_file_names.len() as u64;
```

Comment:
Imagine a directory contains only the following files:
* `part-000000-of-000002`
* `part-000000-of-000003`

The code will see `num_parts = 2` and `total_required_parts = 2` and be happy about the outcome.

Reply:
Ah nice catch! I'll modify